File: native/cocos/core/assets/BitmapFont.cpp
/****************************************************************************
 Copyright (c) 2021-2023 Xiamen Yaji Software Co., Ltd.
 
 http://www.cocos.com
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "BitmapFont.h"
#include "base/Log.h"
#include "gfx-base/GFXDevice.h"
#include "platform/Image.h"
#include "tinyxml2/tinyxml2.h"

namespace cc {

/**
 * BitmapFontFace
 */

BitmapFontFace::BitmapFontFace(Font *font)
: FontFace(font) {
}
void BitmapFontFace::doInit(const FontFaceInfo & /*info*/) {
    const auto &fontPath = _font->getPath();
    const auto &fontData = _font->getData();
    if (fontData.empty()) {
        CC_LOG_ERROR("BitmapFontFace doInit failed: empty font data.");
        return;
    }

    tinyxml2::XMLDocument doc;

    auto error = doc.Parse(reinterpret_cast<const char *>(fontData.data()), fontData.size());
    if (error) {
        CC_LOG_ERROR("BitmapFontFace parse failed.");
        return;
    }

    auto *fontNode = doc.RootElement();
    auto *infoNode = fontNode->FirstChildElement("info");
    _fontSize = infoNode->UnsignedAttribute("size");

    auto *commonNode = fontNode->FirstChildElement("common");
    _lineHeight = commonNode->UnsignedAttribute("lineHeight");
    _textureWidth = commonNode->UnsignedAttribute("scaleW");
    _textureHeight = commonNode->UnsignedAttribute("scaleH");
    uint32_t pages = commonNode->UnsignedAttribute("pages");
    int base = commonNode->IntAttribute("base");

    // load glyphs
    auto *charsNode = fontNode->FirstChildElement("chars");
    auto *charNode = charsNode->FirstChildElement("char");
    while (charNode) {
        FontGlyph glyph;
        uint32_t value{0U};

        uint32_t code = charNode->UnsignedAttribute("id");
        glyph.x = static_cast<int16_t>(charNode->IntAttribute("x"));
        glyph.y = static_cast<int16_t>(charNode->IntAttribute("y"));
        glyph.width = static_cast<uint16_t>(charNode->UnsignedAttribute("width"));
        glyph.height = static_cast<uint16_t>(charNode->UnsignedAttribute("height"));
        glyph.bearingX = static_cast<int16_t>(charNode->IntAttribute("xoffset"));
        glyph.bearingY = static_cast<int16_t>(base - charNode->IntAttribute("yoffset"));
        glyph.advance = static_cast<int16_t>(charNode->IntAttribute("xadvance"));
        glyph.page = charNode->UnsignedAttribute("page");

        _glyphs[code] = glyph;
        charNode = charNode->NextSiblingElement();
    }

    // load textures
    auto *pagesNode = fontNode->FirstChildElement("pages");
    auto *pageNode = pagesNode->FirstChildElement("page");
    _textures.resize(pages, nullptr);

    ccstd::string path = fontPath;
    auto pos = fontPath.rfind('/');

    if (pos == ccstd::string::npos) {
        pos = fontPath.rfind('\\');
    }

    if (pos != ccstd::string::npos) {
        path = fontPath.substr(0, pos + 1);
    }

    while (pageNode) {
        uint32_t id = pageNode->UnsignedAttribute("id");
        ccstd::string file = pageNode->Attribute("file");
        _textures[id] = loadTexture(path + file);

        pageNode = pageNode->NextSiblingElement();
    }

    // load kernings
    auto *kerningsNode = fontNode->FirstChildElement("kernings");
    auto *kerningNode = kerningsNode->FirstChildElement("kerning");
    while (kerningNode) {
        KerningPair pair;
        pair.prevCode = kerningNode->UnsignedAttribute("first");
        pair.nextCode = kerningNode->UnsignedAttribute("second");
        _kernings[pair] = static_cast<float>(kerningNode->IntAttribute("amount"));

        kerningNode = kerningNode->NextSiblingElement();
    }
}

const FontGlyph *BitmapFontFace::getGlyph(uint32_t code) {
    auto iter = _glyphs.find(code);
    if (iter != _glyphs.end()) {
        return &iter->second;
    }

    CC_LOG_WARNING("BitmapFontFace getGlyph failed, character: %u.", code);

    return nullptr;
}

float BitmapFontFace::getKerning(uint32_t prevCode, uint32_t nextCode) {
    if (_kernings.empty()) {
        return 0.0F;
    }

    const auto &iter = _kernings.find({prevCode, nextCode});
    if (iter != _kernings.end()) {
        return iter->second;
    }

    return 0.0F;
}

gfx::Texture *BitmapFontFace::loadTexture(const ccstd::string &path) {
    auto *image = ccnew Image();
    bool success = image->initWithImageFile(path);
    if (!success) {
        CC_LOG_WARNING("BitmapFontFace initWithImageFile failed, path: %s.", path.c_str());
        delete image;
        return nullptr;
    }

    if (image->getRenderFormat() != gfx::Format::R8 && image->getRenderFormat() != gfx::Format::L8) {
        CC_LOG_WARNING("BitmapFontFace loadTexture with invalid format, path: %s.", path.c_str());
        delete image;
        return nullptr;
    }

    auto width = static_cast<uint32_t>(image->getWidth());
    auto height = static_cast<uint32_t>(image->getHeight());

    auto *device = gfx::Device::getInstance();
    auto *texture = device->createTexture({gfx::TextureType::TEX2D,
                                           gfx::TextureUsageBit::SAMPLED | gfx::TextureUsageBit::TRANSFER_DST,
                                           gfx::Format::R8,
                                           width,
                                           height});

    gfx::BufferDataList buffers{image->getData()};
    gfx::BufferTextureCopyList regions = {{0U,
                                           0U,
                                           0U,
                                           {0U, 0U, 0U},
                                           {width, height, 1U},
                                           {0U, 0U, 1U}}};

    device->copyBuffersToTexture(buffers, texture, regions);
    delete image;

    return texture;
}

/**
 * BitmapFont
 */
BitmapFont::BitmapFont(const ccstd::string &path)
: Font(FontType::BITMAP, path) {
}

FontFace *BitmapFont::createFace(const FontFaceInfo &info) {
    auto *face = ccnew BitmapFontFace(this);
    face->doInit(info);

    uint32_t fontSize = face->getFontSize();
    _faces[fontSize] = face;

    return face;
}

} // namespace cc

File: cocos/2d/assets/bitmap-font.ts
/*
 Copyright (c) 2013-2016 Chukong Technologies Inc.
 Copyright (c) 2017-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import { ccclass, type, serializable, editable } from 'cc.decorator';
import { Font } from './font';
import { SpriteFrame } from './sprite-frame';
import { cclegacy, js, warnID } from '../../core';
import { getSymbolCodeAt } from '../utils';
import { IShareLabelInfo } from '../assembler/label/font-utils';
import { TextureBase } from '../../asset/assets/texture-base';

export interface IConfig {
    [key: string]: any;
}

export class FontLetterDefinition {
    public u = 0;
    public v = 0;
    public w = 0;
    public h = 0;
    public offsetX = 0;
    public offsetY = 0;
    public valid = false;
    public xAdvance = 0;
}

export interface ILetterDefinition {
    [key: string]: FontLetterDefinition;
}

export class FontAtlas {
    public letterDefinitions: ILetterDefinition = {};
    public declare texture: TextureBase | null;

    constructor (texture: TextureBase | null) {
        this.texture = texture;
    }

    public addLetterDefinitions (letter: string, letterDefinition: FontLetterDefinition): void {
        this.letterDefinitions[letter] = letterDefinition;
    }

    public cloneLetterDefinition (): ILetterDefinition {
        const copyLetterDefinitions: ILetterDefinition = {};
        for (const key of Object.keys(this.letterDefinitions)) {
            const value = new FontLetterDefinition();
            js.mixin(value, this.letterDefinitions[key]);
            copyLetterDefinitions[key] = value;
        }
        return copyLetterDefinitions;
    }

    public getTexture (): TextureBase | null {
        return this.texture;
    }

    public getLetter (key: string): FontLetterDefinition {
        return this.letterDefinitions[key];
    }

    public getLetterDefinitionForChar (char: string, labelInfo?: IShareLabelInfo): FontLetterDefinition | null {
        const key = getSymbolCodeAt(char, 0);
        const hasKey = Object.prototype.hasOwnProperty.call(this.letterDefinitions, key);
        let letter: FontLetterDefinition | null = null;
        if (hasKey) {
            letter = this.letterDefinitions[key];
        }
        return letter;
    }

    public clear (): void {
        this.letterDefinitions = {};
    }
}

/**
 * @en Class for BitmapFont handling.
 * @zh 位图字体资源类。
 */
@ccclass('cc.BitmapFont')
export class BitmapFont extends Font {
    /**
     * @deprecated since v3.7.0, Useless Code.
     */
    @serializable
    @editable
    public fntDataStr = '';

    /**
     * @en [[SpriteFrame]] of the bitmap font.
     * @zh 位图字体所使用的精灵。
     */
    @type(SpriteFrame)
    public spriteFrame: SpriteFrame | null = null;

    /**
     * @en The font size.
     * @zh 文字尺寸。
     */
    @serializable
    @editable
    public fontSize = -1;

    /**
     * @en Font configuration.
     * @zh 字体配置。
     */
    @serializable
    @editable
    public fntConfig: IConfig | null = null;

    /**
     * @deprecated since v3.7.0, this is an engine private interface that will be removed in the future.
     */
    public declare fontDefDictionary: FontAtlas;

    constructor () {
        super();
    }

    onLoaded (): void {
        const spriteFrame = this.spriteFrame;
        if (!this.fontDefDictionary && spriteFrame) {
            this.fontDefDictionary = new FontAtlas(spriteFrame.texture);
        }

        const fntConfig = this.fntConfig;
        if (!fntConfig) {
            warnID(16376);
            return;
        }

        const fontDict = fntConfig.fontDefDictionary;
        for (const fontDef in fontDict) {
            const info = fontDict[fontDef];
            const letter = new FontLetterDefinition();
            const rect = info.rect;
            letter.offsetX = info.xOffset;
            letter.offsetY = info.yOffset;
            letter.w = rect.width;
            letter.h = rect.height;
            letter.u = rect.x;
            letter.v = rect.y;
            letter.valid = true;
            letter.xAdvance = info.xAdvance;

            this.fontDefDictionary.addLetterDefinitions(fontDef, letter);
        }
    }
}

cclegacy.BitmapFont = BitmapFont;
